rlclientlib/live_model_impl.cc

```diff
@@ -421,8 +421,9 @@ namespace reinforcement_learning {
     i_sender* ranking_data_sender;
 
     // Use the name to create an instance of raw data sender for interactions
+    _configuration.set(name::CONFIG_SECTION, name::INTERACTION);
```

Comment:
Mutating the configuration in order to interpret the values internally prior to calling an `init` function seems like a dangerous pattern to me. What is the rationale here?

Reply:
For this particular one, I did pair with Rajan and he suggested that we should config as a state and mutable not a static settings.

Reply:
@rajan-chari could you explain the rationale then?

Reply:
@rajan-chari @jackgerrits can we please sync tomorrow and close this comment? I have done all the functional testing and ready to merge the PR. This is blocking the merge. If there are any other suggestions, I would prefer we create separate item outside of this PR and address separately since this was already reviewed and approved and is a relatively bigger PR to manage.

Reply:
Normally to construct/initialize objects,  parameters can simply be added to a constructor on init function.  In the case of factory based construction, adding parameters requires changing signature of the factory itself.  The issue there is that this results in create/init methods having to accept a parameter that may or may not be relevant.   This seems acceptable in cases where the extra parameter is a generic.  For example a logger. 

In this case there is a single factory method, that uses this parameter to select different configurations used for construction.  (interaction vs observation).  So adding an extra parameter to config seems like a reasonable choice given the trade off.

I do agree that we are changing a pattern here.  One mitigation might be to move name::CONFIG_SECTION to a different (internal_constants.h ?) file which would allow us to formulate a policy as follows:  
a) Do not modify externally visible name/values passed in via config
b) Adding/Modifying new name/values is allowed as long as these names are in internal_constants.h

Reply:
I like that plan. I think the main thing I was referring to is that we should keep the user inputted configuration in-tact. Adding additional values/arguments is necessary to facilitate the construction like you said.

Moving it to an internal_constants.h hopefully makes it clear to the user and to us working on it that this is not a config value that they should be setting externally but we use it internally for construction of things.

Reply:
Moved the constants to internal_constants.h file.